mirrord/agent/src/steal/http/filter.rs

```diff
@@ -35,6 +37,108 @@ const H2_PREFACE: &[u8; 14] = b"PRI * HTTP/2.0";
 /// version we're dealing with.
 const DEFAULT_HTTP_VERSION_DETECTION_TIMEOUT: Duration = Duration::from_secs(10);
 
+#[derive(Debug)]
+pub(crate) struct HttpHeaderFilter(Regex);
+
+#[derive(Debug)]
+pub(crate) struct HttpPathFilter(Regex);
+
+/// Current supported filtering criterias.
+#[enum_dispatch]
+#[derive(Debug)]
+pub(crate) enum HttpFilter {
+    /// Header based filter.
+    Header(HttpHeaderFilter),
+    /// Path based filter
+    Path(HttpPathFilter),
+}
+
+impl HttpFilter {
+    pub fn new_header_filter(header: Regex) -> Self {
+        Self::Header(HttpHeaderFilter(header))
+    }
+    #[allow(dead_code)]
+    pub fn new_path_filter(path: Regex) -> Self {
+        Self::Path(HttpPathFilter(path))
+    }
+}
+
+/// Struct to hold cache/hold/reuse the attributes that a [`Request`] can be matched upon
+pub(crate) struct RequestMatchCandidate<'a> {
+    /// The original request
+    request: &'a Request<Incoming>,
+    /// Headers in `k: v` format.
+    headers: OnceCell<Vec<String>>,
+}
+
+impl<'a> From<&'a Request<Incoming>> for RequestMatchCandidate<'a> {
+    fn from(request: &'a Request<Incoming>) -> Self {
+        Self {
+            request,
+            headers: OnceCell::new(),
+        }
+    }
+}
+
+impl<'a> RequestMatchCandidate<'a> {
+    /// Returns the headers in `k: v` format.
+    fn headers(&mut self) -> &Vec<String> {
+        self.headers.get_or_init(|| {
+            self.request
+                .headers()
+                .iter()
+                .map(|(header_name, header_value)| {
+                    header_value
+                        .to_str()
+                        .map(|header_value| format!("{header_name}: {header_value}"))
+                })
+                .filter_map(Result::ok)
+                .collect()
+        })
+    }
+
+    /// Returns path of the request.
+    fn path(&self) -> &str {
+        self.request.uri().path()
+    }
+}
+
+/// Trait that needs to be implemented for each [`HttpFilter`] variant.
+#[enum_dispatch(HttpFilter)]
+pub(crate) trait RequestMatch {
+    /// Checks if the request matches the filter.
+    fn matches(&self, request: &mut RequestMatchCandidate<'_>) -> bool;
+}
+
+impl RequestMatch for HttpHeaderFilter {
+    fn matches(&self, request: &mut RequestMatchCandidate<'_>) -> bool {
+        for header in request.headers() {
+            match self.0.is_match(header) {
+                Ok(true) => return true,
+                Ok(false) => continue,
+                Err(err) => {
+                    error!(
+                        "Error while matching header: {header:?}, {:?}, {err:?}",
+                        self.0
+                    );
+                    return false;
+                }
+            }
+        }
+        false
+    }
```

Comment:
```suggestion
request.headers().find_map(|header| self.0.is_match(header).unwrap_or_default()).unwrap_or_default()
```

Something along those lines? (sorry, I can't help myself not suggesting iterators)

Reply:
I think with error handling the loop looks better